### ISSUE TITLE
Several compiler warnings

### ISSUE THREAD

**Issue by mingodad** [smcpeak/smbase#1]
When adding `-Wall -Wextra` we get this:

```
/usr/bin/make -f Makefile
g++ -c -o autofile.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   autofile.cc
g++ -c -o bdffont.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   bdffont.cc
bdffont.cc: In function ‘void parseGlyph(const char*&, BDFFont::Glyph*, const BDFFont&)’:
bdffont.cc:517:39: warning: unused parameter ‘font’ [-Wunused-parameter]
  517 |                        BDFFont const &font)
      |                        ~~~~~~~~~~~~~~~^~~~
g++ -c -o bflatten.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   bflatten.cc
g++ -c -o binary-stdin.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   binary-stdin.cc
binary-stdin.cc: In function ‘int _setmode(int, int)’:
binary-stdin.cc:17:27: warning: unused parameter ‘fd’ [-Wunused-parameter]
   17 |   static int _setmode(int fd, int mode) { return 0; }
      |                       ~~~~^~
binary-stdin.cc:17:35: warning: unused parameter ‘mode’ [-Wunused-parameter]
   17 |   static int _setmode(int fd, int mode) { return 0; }
      |                               ~~~~^~~~
g++ -c -o bitarray.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   bitarray.cc
g++ -c -o boxprint.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   boxprint.cc
g++ -c -o codepoint.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   codepoint.cc
g++ -c -o datablok.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   datablok.cc
In file included from exc.h:15,
                 from datablok.cc:6:
datablok.cc: In member function ‘void DataBlock::init(size_t)’:
datablok.cc:27:25: warning: comparison of unsigned expression >= 0 is always true [-Wtype-limits]
   27 |   xassert(allocatedSize >= 0);
      |           ~~~~~~~~~~~~~~^~~~
xassert.h:25:7: note: in definition of macro ‘xassert’
   25 |     ((cond)? (void)0 : x_assert_fail(#cond, __FILE__, __LINE__))
      |       ^~~~
datablok.cc: In member function ‘void DataBlock::selfCheck() const’:
datablok.cc:52:13: warning: comparison of unsigned expression >= 0 is always true [-Wtype-limits]
   52 |   xassert(0 <= dataLen && dataLen <= allocated);
      |           ~~^~~~~~~~~~
xassert.h:25:7: note: in definition of macro ‘xassert’
   25 |     ((cond)? (void)0 : x_assert_fail(#cond, __FILE__, __LINE__))
      |       ^~~~
datablok.cc: In member function ‘void DataBlock::setDataLen(size_t)’:
datablok.cc:189:13: warning: comparison of unsigned expression >= 0 is always true [-Wtype-limits]
  189 |   xassert(0 <= newLen && newLen <= allocated);
      |           ~~^~~~~~~~~
xassert.h:25:7: note: in definition of macro ‘xassert’
   25 |     ((cond)? (void)0 : x_assert_fail(#cond, __FILE__, __LINE__))
      |       ^~~~
datablok.cc: In member function ‘void DataBlock::setAllocated(size_t)’:
datablok.cc:198:24: warning: comparison of unsigned expression >= 0 is always true [-Wtype-limits]
  198 |   xassert(newAllocated >= 0);
      |           ~~~~~~~~~~~~~^~~~
xassert.h:25:7: note: in definition of macro ‘xassert’
   25 |     ((cond)? (void)0 : x_assert_fail(#cond, __FILE__, __LINE__))
      |       ^~~~
g++ -c -o datetime.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   datetime.cc
g++ -c -o dev-warning.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   dev-warning.cc
g++ -c -o exc.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   exc.cpp
g++ -c -o flatten.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   flatten.cc
g++ -c -o gcc-options.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   gcc-options.cc
g++ -c -o growbuf.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   growbuf.cc
g++ -c -o hashline.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   hashline.cc
In file included from hashline.h:14,
                 from hashline.cc:4:
array.h: In instantiation of ‘void GrowArray<T>::setAllocatedSize(int) [with T = HashLineMap::HashLine]’:
hashline.cc:44:50:   required from here
array.h:204:14: warning: implicitly-declared ‘HashLineMap::HashLine& HashLineMap::HashLine::operator=(const HashLineMap::HashLine&)’ is deprecated [-Wdeprecated-copy]
  204 |       arr[i] = oldArr[i];
      |       ~~~~~~~^~~~~~~~~
In file included from hashline.cc:4:
hashline.h:38:5: note: because ‘HashLineMap::HashLine’ has user-provided ‘HashLineMap::HashLine::HashLine(const HashLineMap::HashLine&)’
   38 |     HashLine(HashLine const &obj)
      |     ^~~~~~~~
In file included from hashline.h:14,
                 from hashline.cc:4:
array.h: In instantiation of ‘void GrowArray<T>::setIndexDoubler(int, const T&) [with T = HashLineMap::HashLine]’:
array.h:299:7:   required from ‘void ArrayStack<T>::push(const T&) [with T = HashLineMap::HashLine]’
hashline.cc:36:56:   required from here
array.h:132:45: warning: implicitly-declared ‘HashLineMap::HashLine& HashLineMap::HashLine::operator=(const HashLineMap::HashLine&)’ is deprecated [-Wdeprecated-copy]
  132 |     { ensureIndexDoubler(index); arr[index] = value; }
      |                                  ~~~~~~~~~~~^~~~~~~
In file included from hashline.cc:4:
hashline.h:38:5: note: because ‘HashLineMap::HashLine’ has user-provided ‘HashLineMap::HashLine::HashLine(const HashLineMap::HashLine&)’
   38 |     HashLine(HashLine const &obj)
      |     ^~~~~~~~
g++ -c -o objcount.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   objcount.cc
g++ -c -o ofstreamts.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   ofstreamts.cc
g++ -c -o parsestring.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   parsestring.cc
g++ -c -o point.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   point.cc
g++ -c -o pprint.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   pprint.cc
g++ -c -o run-process.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   run-process.cc
g++ -c -o sm-file-util.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   sm-file-util.cc
g++ -c -o sm-rc-obj.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   sm-rc-obj.cc
In file included from sm-rc-obj.cc:4:
sm-rc-obj.h: In copy constructor ‘RefCountObject::RefCountObject(const RefCountObject&)’:
sm-rc-obj.h:42:40: warning: unused parameter ‘obj’ [-Wunused-parameter]
   42 |   RefCountObject(RefCountObject const &obj)
      |                  ~~~~~~~~~~~~~~~~~~~~~~^~~
sm-rc-obj.h: In member function ‘RefCountObject& RefCountObject::operator=(const RefCountObject&)’:
sm-rc-obj.h:49:52: warning: unused parameter ‘obj’ [-Wunused-parameter]
   49 |   RefCountObject& operator= (RefCountObject const &obj)
      |                              ~~~~~~~~~~~~~~~~~~~~~~^~~
g++ -c -o sm-to-std-string.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   sm-to-std-string.cc
g++ -c -o smregexp.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   smregexp.cc
g++ -c -o srcloc.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   srcloc.cc
srcloc.cc: In member function ‘int SourceLocManager::File::lineToChar(int)’:
srcloc.cc:371:23: warning: implicitly-declared ‘SourceLocManager::Marker& SourceLocManager::Marker::operator=(const SourceLocManager::Marker&)’ is deprecated [-Wdeprecated-copy]
  371 |     marker = index[low];
      |                       ^
In file included from srcloc.cc:4:
srcloc.h:90:5: note: because ‘SourceLocManager::Marker’ has user-provided ‘SourceLocManager::Marker::Marker(const SourceLocManager::Marker&)’
   90 |     Marker(Marker const &obj)
      |     ^~~~~~
srcloc.cc: In member function ‘void SourceLocManager::File::charToLineCol(int, int&, int&)’:
srcloc.cc:468:23: warning: implicitly-declared ‘SourceLocManager::Marker& SourceLocManager::Marker::operator=(const SourceLocManager::Marker&)’ is deprecated [-Wdeprecated-copy]
  468 |     marker = index[low];
      |                       ^
In file included from srcloc.cc:4:
srcloc.h:90:5: note: because ‘SourceLocManager::Marker’ has user-provided ‘SourceLocManager::Marker::Marker(const SourceLocManager::Marker&)’
   90 |     Marker(Marker const &obj)
      |     ^~~~~~
srcloc.cc: In instantiation of ‘void extractArray(T*&, int&, const ArrayStack<T>&) [with T = SourceLocManager::Marker]’:
srcloc.cc:160:51:   required from here
srcloc.cc:28:14: warning: implicitly-declared ‘SourceLocManager::Marker& SourceLocManager::Marker::operator=(const SourceLocManager::Marker&)’ is deprecated [-Wdeprecated-copy]
   28 |     array[i] = src[i];
      |     ~~~~~~~~~^~~~~~
In file included from srcloc.cc:4:
srcloc.h:90:5: note: because ‘SourceLocManager::Marker’ has user-provided ‘SourceLocManager::Marker::Marker(const SourceLocManager::Marker&)’
   90 |     Marker(Marker const &obj)
      |     ^~~~~~
In file included from srcloc.h:34,
                 from srcloc.cc:4:
array.h: In instantiation of ‘void GrowArray<T>::setIndexDoubler(int, const T&) [with T = SourceLocManager::Marker]’:
array.h:299:7:   required from ‘void ArrayStack<T>::push(const T&) [with T = SourceLocManager::Marker]’
srcloc.cc:79:29:   required from here
array.h:132:45: warning: implicitly-declared ‘SourceLocManager::Marker& SourceLocManager::Marker::operator=(const SourceLocManager::Marker&)’ is deprecated [-Wdeprecated-copy]
  132 |     { ensureIndexDoubler(index); arr[index] = value; }
      |                                  ~~~~~~~~~~~^~~~~~~
In file included from srcloc.cc:4:
srcloc.h:90:5: note: because ‘SourceLocManager::Marker’ has user-provided ‘SourceLocManager::Marker::Marker(const SourceLocManager::Marker&)’
   90 |     Marker(Marker const &obj)
      |     ^~~~~~
In file included from srcloc.h:34,
                 from srcloc.cc:4:
array.h: In instantiation of ‘void GrowArray<T>::setAllocatedSize(int) [with T = SourceLocManager::Marker]’:
array.h:236:3:   required from ‘void GrowArray<T>::eidLoop(int) [with T = SourceLocManager::Marker]’
array.h:128:27:   required from ‘void GrowArray<T>::ensureIndexDoubler(int) [with T = SourceLocManager::Marker]’
array.h:132:7:   required from ‘void GrowArray<T>::setIndexDoubler(int, const T&) [with T = SourceLocManager::Marker]’
array.h:299:7:   required from ‘void ArrayStack<T>::push(const T&) [with T = SourceLocManager::Marker]’
srcloc.cc:79:29:   required from here
array.h:204:14: warning: implicitly-declared ‘SourceLocManager::Marker& SourceLocManager::Marker::operator=(const SourceLocManager::Marker&)’ is deprecated [-Wdeprecated-copy]
  204 |       arr[i] = oldArr[i];
      |       ~~~~~~~^~~~~~~~~
In file included from srcloc.cc:4:
srcloc.h:90:5: note: because ‘SourceLocManager::Marker’ has user-provided ‘SourceLocManager::Marker::Marker(const SourceLocManager::Marker&)’
   90 |     Marker(Marker const &obj)
      |     ^~~~~~
g++ -c -o str.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   str.cc
g++ -c -o strdict.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   strdict.cc
g++ -c -o string-utils.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   string-utils.cc
g++ -c -o stringset.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   stringset.cc
g++ -c -o strtable.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   strtable.cc
g++ -c -o strtokp.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   strtokp.cpp
g++ -c -o strutil.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   strutil.cc
g++ -c -o svdict.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   svdict.cc
g++ -c -o syserr.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   syserr.cpp
g++ -c -o trace.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   trace.cc
g++ -c -o trdelete.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   trdelete.cc
g++ -c -o tree-print.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   tree-print.cc
gcc -c -o unixutil.o -MMD  -g -O2 -Wall -Wextra -std=c99   unixutil.c
g++ -c -o vdtllist.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   vdtllist.cc
g++ -c -o voidlist.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   voidlist.cc
g++ -c -o vptrmap.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   vptrmap.cc
g++ -c -o warn.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   warn.cpp
rm -f libsmbase.a
ar -r libsmbase.a autofile.o bdffont.o bflatten.o binary-stdin.o bit2d.o bitarray.o boxprint.o breaker.o codepoint.o crc.o cycles.o d2vector.o datablok.o datetime.o dev-warning.o exc.o flatten.o functional-set.o gcc-options.o gprintf.o growbuf.o hashline.o hashtbl.o missing.o mypopen.o mysig.o nonport.o objcount.o ofstreamts.o parsestring.o point.o pprint.o refct-serf.o run-process.o sm-compare.o sm-file-util.o sm-rc-obj.o sm-to-std-string.o smregexp.o srcloc.o str.o strdict.o strhash.o string-utils.o stringset.o strtable.o strtokp.o strutil.o svdict.o syserr.o trace.o trdelete.o tree-print.o unixutil.o vdtllist.o voidlist.o vptrmap.o warn.o
ar: creating libsmbase.a
ranlib libsmbase.a
g++ -c -o array-test.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   array-test.cc
g++ -c -o astlist-test.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   astlist-test.cc
g++ -c -o bflatten-test.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   bflatten-test.cc
g++ -c -o datablok-test.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   datablok-test.cc
g++ -c -o dict-test.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   dict-test.cc
In file included from dict-test.cc:8:
strintdict.h: In member function ‘void StringIntDict::foreach(StringIntDict::ForeachFn, void*) const’:
strintdict.h:178:47: warning: cast between incompatible function types from ‘StringIntDict::ForeachFn’ {aka ‘bool (*)(const string&, long int, void*)’} to ‘StringVoidDict::ForeachFn’ {aka ‘bool (*)(const string&, void*, void*)’} [-Wcast-function-type]
  178 |     { dict.foreach((StringVoidDict::ForeachFn)func, extra); }
      |                                               ^~~~
g++ -c -o functional-set-test.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   functional-set-test.cc
In file included from sm-rc-ptr.h:7,
                 from functional-set.h:9,
                 from functional-set-test.cc:4:
sm-rc-obj.h: In copy constructor ‘RefCountObject::RefCountObject(const RefCountObject&)’:
sm-rc-obj.h:42:40: warning: unused parameter ‘obj’ [-Wunused-parameter]
   42 |   RefCountObject(RefCountObject const &obj)
      |                  ~~~~~~~~~~~~~~~~~~~~~~^~~
sm-rc-obj.h: In member function ‘RefCountObject& RefCountObject::operator=(const RefCountObject&)’:
sm-rc-obj.h:49:52: warning: unused parameter ‘obj’ [-Wunused-parameter]
   49 |   RefCountObject& operator= (RefCountObject const &obj)
      |                              ~~~~~~~~~~~~~~~~~~~~~~^~~
g++ -c -o gcc-options-test.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   gcc-options-test.cc
g++ -c -o map-utils-test.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   map-utils-test.cc
g++ -c -o overflow-test.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   overflow-test.cc
g++ -c -o parsestring-test.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   parsestring-test.cc
g++ -c -o sm-rc-ptr-test.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   sm-rc-ptr-test.cc
In file included from sm-rc-ptr.h:7,
                 from sm-rc-ptr-test.cc:4:
sm-rc-obj.h: In copy constructor ‘RefCountObject::RefCountObject(const RefCountObject&)’:
sm-rc-obj.h:42:40: warning: unused parameter ‘obj’ [-Wunused-parameter]
   42 |   RefCountObject(RefCountObject const &obj)
      |                  ~~~~~~~~~~~~~~~~~~~~~~^~~
sm-rc-obj.h: In member function ‘RefCountObject& RefCountObject::operator=(const RefCountObject&)’:
sm-rc-obj.h:49:52: warning: unused parameter ‘obj’ [-Wunused-parameter]
   49 |   RefCountObject& operator= (RefCountObject const &obj)
      |                              ~~~~~~~~~~~~~~~~~~~~~~^~~
sm-rc-ptr-test.cc: In copy constructor ‘Foo::Foo(const Foo&)’:
sm-rc-ptr-test.cc:28:3: warning: base class ‘class RefCountObject’ should be explicitly initialized in the copy constructor [-Wextra]
   28 |   Foo(Foo const &obj)
      |   ^~~
g++ -c -o string-utils-test.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   string-utils-test.cc
g++ -c -o vector-utils-test.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   vector-utils-test.cc
g++ -c -o unit-tests.o -MMD  -g -O2 -Wall -Wextra  -std=c++11   unit-tests.cc
g++ -o unit-tests.exe -g -O2 -Wall -Wextra  -std=c++11     array-test.o astlist-test.o bflatten-test.o datablok-test.o dict-test.o functional-set-test.o gcc-options-test.o map-utils-test.o overflow-test.o parsestring-test.o sm-rc-ptr-test.o string-utils-test.o vector-utils-test.o unit-tests.o libsmbase.a 
g++ -o binary-stdin-test.exe -g -O2 -Wall -Wextra  -std=c++11    binary-stdin-test.cc libsmbase.a 

BUILD SUCCESSFUL (total time: 32s)
```

**Comment by smcpeak** [comment]
Thanks for the report.  I've resolved as much of this as I think makes sense as of commit 3ee3c13e95, so closing this.

I chose not to do anything about the comparisons of unsigned against 0, instead passing `-Wno-type-limits` during my testing.  These checks are inside assertions, where obviously I do not expect them to be false.  I want the assertions to be as complete as possible, even when some aspects are guaranteed by signedness, as a form of documentation.  Ideally, GCC would not report when the false branch would cause an assertion to fail.  (I experimented with using `static_assert` and `std::is_unsigned`, but ultimately decided that was not an improvement.)

I also chose not to do anything about the function cast in `strintdict.h`, instead passing `-Wno-cast-function-type` and adding an explanatory comment about why.  In this case, GCC is probably technically right, but in the absence of a problem other than the warning itself, I think it's not worth fixing.  That code is quite old, and in new code I use the standard C++ containers instead.
